proposals/progress_output.md

```diff
@@ -0,0 +1,177 @@
+# Proposal: Provide feedback on push progress
+
+Relevant issues: [#806](https://github.com/GoogleContainerTools/jib/issues/806), [#1251](https://github.com/GoogleContainerTools/jib/issues/1251)
+
+## Motivation
+
+Currently, Jib lacks any progress feedback for layer pushes. In cases where layers take a long time to push, this lack of progress indication:
+
+- makes it hard to estimate how long the build will take
+- makes it unclear as to whether the build is stuck or will complete in some reasonable amount of time
+- may deter first-time users from continuing to use Jib 
+
+This is especially prevalent for first-time users whose first builds would need to push all the layers, including the large base image layers.
+
+## Current output
+
+An example of the current output (as of versoin `0.10.0`) looks like:
+
+```
+Containerizing application to <image>...
+warning: Base image 'gcr.io/distroless/java' does not use a specific image digest - build may not be reproducible
+Retrieving registry credentials for gcr.io...
+Getting base image gcr.io/distroless/java...
+Building dependencies layer...
+Building resources layer...
+Building classes layer...
+
+Container entrypoint set to [java, -cp, /app/resources:/app/classes:/app/libs/*, <main class>]
+Finalizing...
+
+Built and pushed image as <image>
+```
+
+Note that the build would pause on `Finalizing...` until all layers are finished pushing.
+
+## Goal
+
+To display fine-grained progress feedback during the build, but not clutter the log messages.
+
+## Design considerations
+
+Since Jib builds/pulls and pushes each layer independently, synchronized progress feedback (completion over all layers, such as a progress bar) could potentially limit the concurrency of the build process - but this might be negligible.
+
+The available information for each layer include:
+
+- type (classes, resources, dependencies, base image layer, etc.)
+- digest
+- size
+
+The non-available information for each layer include:
+
+- sizes of other layers (eg. other layers may not have finished building)
+- number of other layers (eg. base image layers have not been resolved before starting to push an application layer)
+
+Therefore, the actions we can perform for progress feedback include:
+
+- monitor the number of bytes sent (and percentage completion)
+
+And the actions we cannot perform include:
+
+- know ahead of time how many layers will be pulled/pushed
+
+## Proposal
+
+Display an overall progress bar along with the tasks currently being executed.
```

Comment:
How will this work at different logging levels?

Reply:
I think we can disable the progress bar for logging levels debug and higher since there will be many log messages to indicate progress anyways, but the progress bar itself should be implemented such that it does not interfere with the other log messages Jib emits.